### PR TITLE
Http client pool improvements

### LINKED ISSUE #2787: HttpClient pool improvements
A few improvements done to the HttpClient and the pool:

- the client suffers from races when a connection is closed while being delivered to the client request (see https://github.com/eclipse-vertx/vert.x/issues/2773)
- TCP idle timeout now is not handled anymore with pool events but instead when a TCP idle event happens, it is ignored when the connection is in the pool
- the activate/passivate connection has been removed (since it's not used anymore for TCP idle)
- the coupling of the pool to channel has been removed as it was not necessary
- 

The main change is that now the pool methods update the pool state and the corresponding events (deliver a connection to a waiter, connect to an endpoint, keep the waiter list bounded) are handled on the event-loop thread to avoid races. To accomodate this, now the pool has a context instead of having a context per connection.

### PR DESCRIPTION
**Pull request by vietj**
Fixes #2787

### CODE REVIEW

**Comment by vietj** [comment]
note : the PR should be merged with a squashed commit but I've let the original commits to ease the review

**Comment by vietj** [comment]
@tsegismont can you review this PR ?

File: src/main/java/io/vertx/core/http/impl/pool/Pool.java [hunk 1]
```diff
@@ -25,43 +25,58 @@
...
  * keep alive timeout.
  *
- * When a waiter asks for a connection, it is either added to the queue (when it's not empty) or attempted to be
- * served (from the pool or by creating a new connection) or failed. The {@link #waitersCount} is the number
- * of total waiters (the waiters in {@link #waitersQueue} but also the inflight) so we know if we can close the pool
- * or not. The {@link #waitersCount} is incremented when a waiter wants to acquire a connection succesfully (i.e
- * it is either added to the queue or served from the pool) and decremented when the it gets a reply (either with
- * a connection or with a failure).
+ * <h3>Acquiring a connection</h3>
+ * When a waiter wants to acquire a connection it is either added to the {@link #waitersQueue} and the request
```
Comment (tsegismont) [inline src/main/java/io/vertx/core/http/impl/pool/Pool.java]: You can remove "is either". Now it's always added to the waiter queue.

File: src/main/java/io/vertx/core/http/impl/pool/Pool.java [hunk 2]
```diff
@@ -25,43 +25,58 @@
...
+ * will be handled by the pool asynchronously.
+ * <ul>
+ *   <li>when there is an available pooled connection, this connection is delivered to the waiter and it is removed
+ *   from the queue</li>
+ *   <li>when there is no available connection a connection is created when {@link #weight}{@code <}{@link #maxWeight}</li>
+ *   <li>when the max number of waiters is reached, the request is failed</li>
+ *   <li>otherwise the waiter remains in the queue until progress can be done (i.e a connection is recycled, etc...)</li>
+ * </ul>
+ * When a waiter is satisfied, to avoid races the waiter is guaranteed to read the correct connection state.
```
Comment (tsegismont) [inline src/main/java/io/vertx/core/http/impl/pool/Pool.java]: Could you rephrase this one? It's not clear what you mean.

File: src/main/java/io/vertx/core/http/impl/pool/Pool.java [hunk 3]
```diff
@@ -346,6 +350,7 @@ private synchronized void setConcurrency(Holder holder, long concurrency) {
       throw new IllegalArgumentException("Cannot set a negative concurrency value");
     }
     if (holder.removed) {
+      assert false : "Cannot recycle removed holder";
```
Comment (tsegismont) [inline src/main/java/io/vertx/core/http/impl/pool/Pool.java]: Do we want to keep this assertion? We don't use them afaik

File: src/main/java/io/vertx/core/http/impl/ConnectionManager.java [hunk 4]
```diff
@@ -124,16 +124,16 @@ void getConnection(ContextInternal ctx, String peerHost, boolean ssl, int port,
...
-          connectionMap::remove,
+          conn -> connectionMap.put(conn.channel(), conn),
+          conn -> connectionMap.remove(conn.channel(), conn),
           false);
         return new Endpoint(pool, metric);
       });
       Object metric;
       if (metrics != null) {
-        metric = metrics.enqueueRequest(endpoint.metric); // Metrics might be enqueued several times if getConnection returns false
+        metric = metrics.enqueueRequest(endpoint.metric);
```
Comment (tsegismont) [inline src/main/java/io/vertx/core/http/impl/ConnectionManager.java]: The comment was added in one the first commits. Was it before the logic was refactored and now the risk has been mitigated? I'm asking because I'm not sure why this comment was added and now removed without the surrounding code being modified.

**Comment by tsegismont** [comment]
@vietj looks good to me. A few minor concerns.

File: src/main/java/io/vertx/core/http/impl/Http2ClientConnection.java [hunk 5]
```diff
@@ -466,7 +466,15 @@ public NetSocket createNetSocket() {
       return conn.toNetSocket(this);
     }
   }
-  
+
+  @Override
+  protected void handleIdle() {
+    if (streams.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (vietj) [reply src/main/java/io/vertx/core/http/impl/Http2ClientConnection.java]: I think it should indeed, good catch!

File: src/main/java/io/vertx/core/http/impl/pool/Pool.java (hunk 3 above)
Comment (vietj) [reply src/main/java/io/vertx/core/http/impl/pool/Pool.java]: it is actually used in tests (enabled when running tests) which is very useful

File: src/main/java/io/vertx/core/http/impl/ConnectionManager.java (hunk 4 above)
Comment (vietj) [reply src/main/java/io/vertx/core/http/impl/ConnectionManager.java]: it was before

File: src/main/java/io/vertx/core/http/impl/pool/Pool.java (hunk 3 above)
Comment (vietj) [reply src/main/java/io/vertx/core/http/impl/pool/Pool.java]: I mean that it can catch more bugs

Comment (vietj) [reply src/main/java/io/vertx/core/http/impl/pool/Pool.java]: (and it did :-) )
